src/to_wkt/geo_trait_impl.rs

```diff
@@ -0,0 +1,446 @@
+        }
+
+        for polygon in polygons {
+            f.write_str("),(")?;
+
+            add_coord_sequence(polygon.exterior().unwrap().coords(), f, size)?;
+            for interior in polygon.interiors() {
+                f.write_char(',')?;
+                add_coord_sequence(interior.coords(), f, size)?;
+            }
+        }
+
+        f.write_str("))")?;
+    } else {
+        f.write_str(" EMPTY")?;
+    };
+
+    Ok(())
+}
+
+/// Create geometry to WKT representation.
+
+pub fn geometry_to_wkt<T: CoordNum + WktNum + fmt::Display, G: GeometryTrait<T = T>, W: Write>(
+    geometry: &G,
+    f: &mut W,
+) -> Result<(), std::fmt::Error> {
+    match geometry.as_type() {
+        geo_traits::GeometryType::Point(point) => point_to_wkt(point, f),
+        geo_traits::GeometryType::LineString(linestring) => linestring_to_wkt(linestring, f),
+        geo_traits::GeometryType::Polygon(polygon) => polygon_to_wkt(polygon, f),
+        geo_traits::GeometryType::MultiPoint(multi_point) => multi_point_to_wkt(multi_point, f),
+        geo_traits::GeometryType::MultiLineString(mls) => multi_linestring_to_wkt(mls, f),
+        geo_traits::GeometryType::MultiPolygon(multi_polygon) => {
+            multi_polygon_to_wkt(multi_polygon, f)
+        }
+        geo_traits::GeometryType::GeometryCollection(gc) => geometry_collection_to_wkt(gc, f),
+        geo_traits::GeometryType::Rect(rect) => rect_to_wkt(rect, f),
+        geo_traits::GeometryType::Triangle(triangle) => triangle_to_wkt(triangle, f),
+        geo_traits::GeometryType::Line(line) => line_to_wkt(line, f),
+    }
+}
+
+pub fn geometry_collection_to_wkt<
+    T: CoordNum + WktNum + fmt::Display,
+    G: GeometryCollectionTrait<T = T>,
+    W: Write,
+>(
+    gc: &G,
+    f: &mut W,
+) -> Result<(), std::fmt::Error> {
+    let dim = gc.dim();
+    // Write prefix
+    match dim {
+        geo_traits::Dimensions::Xy => f.write_str("GEOMETRYCOLLECTION"),
+        geo_traits::Dimensions::Xyz => f.write_str("GEOMETRYCOLLECTION Z"),
+        geo_traits::Dimensions::Xym => f.write_str("GEOMETRYCOLLECTION M"),
+        geo_traits::Dimensions::Xyzm => f.write_str("GEOMETRYCOLLECTION ZM"),
+        geo_traits::Dimensions::Unknown(_) => todo!(),
+    }?;
+    let mut geometries = gc.geometries();
+
+    if let Some(first_geometry) = geometries.next() {
+        f.write_str("(")?;
+
+        geometry_to_wkt(&first_geometry, f)?;
+        for geom in geometries {
+            f.write_char(',')?;
+            geometry_to_wkt(&geom, f)?;
+        }
+
+        f.write_char(')')?;
+    } else {
+        f.write_str(" EMPTY")?;
+    }
+    Ok(())
+}
+
+fn rect_to_polygon<T: CoordNum + WktNum + fmt::Display, G: RectTrait<T = T>>(
+    rect: &G,
+) -> Polygon<T> {
+    let min_coord = rect.min();
+    let max_coord = rect.max();
+
+    // Note: Even if the rect has more than 2 dimensions, we omit the other dimensions when
+    // converting to a Polygon.
```

Comment:
`Triangle` and `Line` store _all_ coordinates, just with the simplifying case that they only have a specific number of points. `Rect` doesn't do that, or else `Rect` would store _4_ coordinates, not 2. We need to do some sort of conversion from the `min` and `max` coordinates to a continuous ring.

It's obvious how to do this for 2D input, but less clear how to do this for 3D or 4D. I'd suggest that we should error for 3D/4D, as the user can choose how to transform a 3D or 4D Rect.

Or perhaps we could even change the `RectTrait` so that it doesn't have the same dimensionality as the others. Because `Rect` means "axis aligned bounding box" and when you have 3 or 4 axes, it's less clear how that maps to simple-features.

Reply:
oh geeze, yeah that's a mess.

Given that we're only handling 2d, can we at least avoid the vec allocation in that case?

As for erroring on 3-d/4-d, that seems reasonable, assuming you mean a runtime error.

Reply:
Yeah, we can avoid the vec allocation, just makes the code slightly more complicated.

Also, what should we do with the error? Currently it returns `std::fmt::Error` and there's no error type in this crate.

Reply:
Feel free to propose something.

Reply:
I added an `Error` enum and now error on non-2d Rect input.